src/audio/module_adapter/module/memory-heap.c

```diff
@@ -0,0 +1,484 @@
+	list_item_prepend(&container->list, &res->res_list);
+
+	return ptr;
+}
+EXPORT_SYMBOL(mod_fast_get);
+#endif
+
+static int free_contents(struct processing_module *mod, struct module_resource *container)
+{
+	struct module_resources *res = &mod->priv.resources;
+
+	switch (container->type) {
+	case MOD_RES_HEAP:
+		sof_heap_free(res->heap, container->ptr);
+		res->heap_usage -= container->size;
+		return 0;
+#if CONFIG_COMP_BLOB
+	case MOD_RES_BLOB_HANDLER:
+		comp_data_blob_handler_free(container->bhp);
+		return 0;
+#endif
+#if CONFIG_FAST_GET
+	case MOD_RES_FAST_GET:
+		fast_put(res->heap, container->sram_ptr);
+		return 0;
+#endif
+	default:
+		comp_err(mod->dev, "Unknown resource type: %d", container->type);
+	}
+	return -EINVAL;
+}
+
+/**
+ * Frees the memory block removes it from module's book keeping.
+ * @param mod	Pointer to module this memory block was allocated for.
+ * @param ptr	Pointer to the memory block.
+ */
+int mod_free(struct processing_module *mod, const void *ptr)
+{
+	struct module_resources *res = &mod->priv.resources;
+	struct module_resource *container;
+	struct list_item *res_list;
+
+	MEM_API_CHECK_THREAD(res);
+	if (!ptr)
+		return 0;
+
+	/* Find which container keeps this memory */
+	list_for_item(res_list, &res->res_list) {
+		container = container_of(res_list, struct module_resource, list);
+		if (container->ptr == ptr) {
+			int ret = free_contents(mod, container);
+
+			list_item_del(&container->list);
+			container_put(mod, container);
+			return ret;
+		}
+	}
+
+	comp_err(mod->dev, "error: could not find memory pointed by %p", ptr);
+
+	return -EINVAL;
+}
+EXPORT_SYMBOL(mod_free);
+
+/**
+ * Frees all the resources registered for this module
+ * @param mod	Pointer to module that should have its resource freed.
+ *
+ * This function is called automatically when the module is unloaded.
+ */
+void mod_free_all(struct processing_module *mod)
+{
+	struct module_resources *res = &mod->priv.resources;
+	struct k_heap *mod_heap = res->heap;
+	struct list_item *list;
+	struct list_item *_list;
+
+	MEM_API_CHECK_THREAD(res);
+	/* Free all contents found in used containers */
+	list_for_item(list, &res->res_list) {
+		struct module_resource *container =
+			container_of(list, struct module_resource, list);
+
+		free_contents(mod, container);
+	}
+
+	/*
+	 * We do not need to remove the containers from res_list in
+	 * the loop above or go through free_cont_list as all the
+	 * containers are anyway freed in the loop below, and the list
+	 * heads are reinitialized when mod_resource_init() is called.
+	 */
+	list_for_item_safe(list, _list, &res->cont_chunk_list) {
+		struct container_chunk *chunk =
+			container_of(list, struct container_chunk, chunk_list);
+
+		list_item_del(&chunk->chunk_list);
+		sof_heap_free(mod_heap, chunk);
+	}
+
+	/* Make sure resource lists and accounting are reset */
+	mod_resource_init(mod);
+}
+EXPORT_SYMBOL(mod_free_all);
+
+int module_free(struct processing_module *mod)
```

Comment:
This function does not appear to be related to memory. Can you explain why it was moved?

Reply:
It does some memory freeing, not quite the opposite of init() which is common.